### ISSUE TITLE
Unexpected panic when updating a resource

### ISSUE THREAD

**Issue by chrsmith** [pulumi/pulumi#3953]
I have a stack using Azure Container Instances that's more or less working as expected.

If I make a change to a resource's definition and then run `pulumi up --skip-preview`, `pulumi` will panic with "Expected diff to not require deletion or replacement during update".

This could be specific to the Azure resource provider, but I am guessing it has something do with how we are computing a diff and passing it around, and the fact this is removing an element of an array inside of type, which is itself an element of an array, which is then passed as an input property. (And there are enough transitions there that something might go awry.)

**Full stack trace**
```
    panic: fatal: An assertion has failed: Expected diff to not require deletion or replacement during Update of urn:pulumi:seablock::factorio-server::azure:containerservice/group:Group::factorio-server
    goroutine 94 [running]:
    github.com/pulumi/pulumi/pkg/util/contract.failfast(...)
        /home/travis/gopath/pkg/mod/github.com/pulumi/pulumi@v1.9.1/pkg/util/contract/failfast.go:23
    github.com/pulumi/pulumi/pkg/util/contract.Assertf(0xc000a5e500, 0x4e172f5, 0x48, 0xc00091d668, 0x1, 0x1)
        /home/travis/gopath/pkg/mod/github.com/pulumi/pulumi@v1.9.1/pkg/util/contract/assert.go:33 +0x197
    github.com/pulumi/pulumi-terraform-bridge/pkg/tfbridge.(*Provider).Update(0xc000575680, 0x54fbca0, 0xc000a1ebd0, 0xc00050a690, 0xc000575680, 0x47e4501, 0xc000abb940)
        /home/travis/gopath/pkg/mod/github.com/pulumi/pulumi-terraform-bridge@v1.6.6/pkg/tfbridge/provider.go:805 +0x81d
    github.com/pulumi/pulumi/sdk/proto/go._ResourceProvider_Update_Handler.func1(0x54fbca0, 0xc000a1ebd0, 0x4b297a0, 0xc00050a690, 0x4b24ca0, 0x8008420, 0x54fbca0, 0xc000a1ebd0)
        /home/travis/gopath/pkg/mod/github.com/pulumi/pulumi@v1.9.1/sdk/proto/go/provider.pb.go:1651 +0x86
    github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc.OpenTracingServerInterceptor.func1(0x54fbca0, 0xc000688cc0, 0x4b297a0, 0xc00050a690, 0xc000789800, 0xc000789820, 0x0, 0x0, 0x546b3e0, 0xc000428b60)
        /home/travis/gopath/pkg/mod/github.com/grpc-ecosystem/grpc-opentracing@v0.0.0-20171105060200-01f8541d5372/go/otgrpc/server.go:61 +0x36e
    github.com/pulumi/pulumi/sdk/proto/go._ResourceProvider_Update_Handler(0x4b9eba0, 0xc000575680, 0x54fbca0, 0xc000688cc0, 0xc001007800, 0xc00011c8c0, 0x54fbca0, 0xc000688cc0, 0xc0000e8400, 0x8a2)
        /home/travis/gopath/pkg/mod/github.com/pulumi/pulumi@v1.9.1/sdk/proto/go/provider.pb.go:1653 +0x14b
    google.golang.org/grpc.(*Server).processUnaryRPC(0xc0000cc2c0, 0x554dca0, 0xc0000a5980, 0xc000bd0800, 0xc000a123f0, 0x7fdb220, 0x0, 0x0, 0x0)
        /home/travis/gopath/pkg/mod/google.golang.org/grpc@v1.24.0/server.go:995 +0x460
    google.golang.org/grpc.(*Server).handleStream(0xc0000cc2c0, 0x554dca0, 0xc0000a5980, 0xc000bd0800, 0x0)
        /home/travis/gopath/pkg/mod/google.golang.org/grpc@v1.24.0/server.go:1275 +0xd97
    google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc0003d7610, 0xc0000cc2c0, 0x554dca0, 0xc0000a5980, 0xc000bd0800)
        /home/travis/gopath/pkg/mod/google.golang.org/grpc@v1.24.0/server.go:710 +0xbb
    created by google.golang.org/grpc.(*Server).serveStreams.func1
        /home/travis/gopath/pkg/mod/google.golang.org/grpc@v1.24.0/server.go:708 +0xa1
```

**Snippet**
This is a subset of the full Pulumi program. `pulumi up --skip-preview` works just fine, but if I remove the highlighted lines (removing a port to be exposed from the container), it will trigger the panic.

```diff
const containerImage = new azure.containerservice.Group("factorio-server", {
    containers: [{
        name: "factorio",

        image: factorioGameImage.imageName,

        // Machine resources.
        memory: 1,
        cpu: 1,

        // Ports to expose.
        ports: [
            // Game server (required).
            {
                port: 34197,
                protocol: "UDP",
            },
-            // RCON for remote management (optional).
-            {
-                port: 27015,
-                protocol: "TCP"
-            },
        ],
    }],

    // Provide credentials to fetch the image from our container repository.
    imageRegistryCredentials: [
        {
            server: registry.loginServer,
            username: registry.adminUsername,
            password: registry.adminPassword,
        }
    ],

    ipAddressType: "public",
    osType: "linux",

    dnsNameLabel: `factorio-${pulumi.getStack()}`,

    resourceGroupName: resourceGroup.name,
    location: resourceGroup.location,
});
```

This is from using `v1.10.1`.

**Comment by tenwit** [comment]
I can reproduce this without changing the array attribute. It does seem to require having an array attribute (in my case, it's the vpcSecurityGroupIds of an AWS EC2 instance), but I can leave those along and just add a tag, and I'm seeing the same thing. Currently, this is preventing me from updating EC2 instances, I have to destroy and re-created.

**Comment by lukehoban** [comment]
@tenwit Can you share any more details on the repro for your case?

**Comment by joeduffy** [comment]
I've reproduced this, confirmed the root cause is the same as https://github.com/pulumi/pulumi/issues/5637#issuecomment-738015872, and that the hypothetical fix https://github.com/pulumi/pulumi-terraform-bridge/commit/24364cc31ea77641393729b6b97af6ba8fac460b addresses the panic.

**Comment by komalali** [comment]
This issue has been addressed, pulumi-aws v3.19.1 includes the fix.
